src/Features/CSharp/Portable/UseRecursivePatterns/AnalyzedNode.cs

```diff
@@ -0,0 +1,408 @@
+                Pair pair => ImmutableArray.Create(pair.Input, pair.Pattern),
+                _ => ImmutableArray<AnalyzedNode>.Empty,
+            };
+        }
+
+        internal sealed class True : AnalyzedNode
+        {
+            public static readonly AnalyzedNode Instance = new True();
+            private True() { }
+        }
+
+        internal sealed class False : AnalyzedNode
+        {
+            public static readonly AnalyzedNode Instance = new False();
+            private False() { }
+        }
+
+        internal sealed class Not : AnalyzedNode
+        {
+            public readonly AnalyzedNode Operand;
+            private Not(AnalyzedNode operand) => Operand = operand;
+
+            [return: NotNullIfNotNull("operand")]
+            public static AnalyzedNode? Create(AnalyzedNode? operand)
+            {
+                Debug.Assert(!(operand is OperationEvaluation), "!(operand is OperationEvaluation)");
+                return operand switch
+                {
+                    null => null,
+                    True _ => False.Instance,
+                    False _ => True.Instance,
+                    Not p => p.Operand,
+                    Relational p => new Relational(p.Input, Negate(p.OperatorKind), p.Value),
+                    _ => new Not(operand)
+                };
+
+                static BinaryOperatorKind Negate(BinaryOperatorKind kind)
+                {
+                    return kind switch
+                    {
+                        BinaryOperatorKind.LessThan => BinaryOperatorKind.GreaterThanOrEqual,
+                        BinaryOperatorKind.GreaterThan => BinaryOperatorKind.LessThanOrEqual,
+                        BinaryOperatorKind.LessThanOrEqual => BinaryOperatorKind.GreaterThan,
+                        BinaryOperatorKind.GreaterThanOrEqual => BinaryOperatorKind.LessThan,
+                        var v => throw ExceptionUtilities.UnexpectedValue(v)
+                    };
+                }
+            }
+        }
+
+        internal abstract class Sequence : AnalyzedNode
+        {
+            public readonly ImmutableArray<AnalyzedNode> Nodes;
+            protected Sequence(ImmutableArray<AnalyzedNode> nodes) => Nodes = nodes;
+
+            public abstract AnalyzedNode Update(ArrayBuilder<AnalyzedNode> nodes);
+            public abstract AnalyzedNode Negate(ArrayBuilder<AnalyzedNode> nodes);
+
+            public static AnalyzedNode Create(bool disjunctive, ArrayBuilder<AnalyzedNode> builder)
+                => disjunctive ? OrSequence.Create(builder) : AndSequence.Create(builder);
+        }
+
+        internal sealed class AndSequence : Sequence
+        {
+            public AndSequence(ImmutableArray<AnalyzedNode> nodes) : base(nodes) { }
+
+            public override AnalyzedNode Update(ArrayBuilder<AnalyzedNode> nodes) => Create(nodes);
+            public override AnalyzedNode Negate(ArrayBuilder<AnalyzedNode> nodes) => OrSequence.Create(nodes);
```

Comment:
doc why this is correctr.

Reply:
Perhaps `Negate` is not the best name. we're just creating a node opposite to self. AndSequence creates OrSequence.